app/models/accounting/interest_calculator.rb

```diff
@@ -60,75 +60,101 @@ def initialize(loan)
     def recalculate
       prev_tx = nil
 
-      ensure_interest_transactions
-
-      transactions.each do |tx|
-        case tx.loan_transaction_type_value
-        when "interest"
-          tx.amount = accrued_interest(prev_tx, tx)
-          line_item_for(tx, int_rcv_acct).assign_attributes(
-            qb_line_id: 0,
-            posting_type: "Debit",
-            amount: tx.amount
-          )
-          line_item_for(tx, int_inc_acct).assign_attributes(
-            qb_line_id: 1,
-            posting_type: "Credit",
-            amount: tx.amount
-          )
-
-        when "disbursement"
-          line_item_for(tx, prin_acct).assign_attributes(
-            qb_line_id: 0,
-            posting_type: "Debit",
-            amount: tx.amount
-          )
-          line_item_for(tx, tx.account).assign_attributes(
-            qb_line_id: 1,
-            posting_type: "Credit",
-            amount: tx.amount
-          )
-
-        when "repayment"
-          int_part = [tx.amount, prev_tx.try(:interest_balance) || 0].min
-          line_item_for(tx, tx.account).assign_attributes(
-            qb_line_id: 0,
-            posting_type: "Debit",
-            amount: tx.amount
-          )
-          line_item_for(tx, prin_acct).assign_attributes(
-            qb_line_id: 1,
-            posting_type: "Credit",
-            amount: tx.amount - int_part
-          )
-          line_item_for(tx, int_rcv_acct).assign_attributes(
-            qb_line_id: 2,
-            posting_type: "Credit",
-            amount: int_part
-          )
-        end
-
-        # Since we may have just adjusted line items upon which the change_in_principal and
-        # change_in_interest depend, it is important that we recalculate balances now.
-        tx.calculate_balances(prev_tx: prev_tx)
+      txns_by_date = transactions.group_by(&:txn_date)
+      first_date = transactions.first&.txn_date
+      last_date = loan.status_value == 'active' ? Date.today : transactions.last&.txn_date
 
-        # Before we save, check if the transaction's line items have changed and
-        # set the needs_qb_push flag. We ignore changes to the transaction's balances since these
-        # don't need to get pushed. Therefore we don't need to check changes on the transaction
-        # object itself, just the line items.
-        #
-        # Note that if the flag is already set we leave it as true even if no changes
-        # occurred. The transaction may have changed earlier (e.g. via the UI) and may need a push
-        # even if we don't change anything here.
-        tx.needs_qb_push = tx.needs_qb_push || tx.line_items.any?(&:type_or_amt_changed?)
+      txn_dates = txns_by_date.keys
+      month_boundaries = month_boundaries(first_date, last_date)
```

Comment:
I might rename this variable, I try to avoid using temporary variable names that are the same as method names because it's harder when you start trying to refactor things (find/replace will catch false positives)